File: leetcode-solutions/go/141. Linked List Cycle.go
// https://leetcode.com/problems/linked-list-cycle/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func hasCycle(head *ListNode) bool {
	slow, fast := head, head
	
	for ; fast != nil; {
			slow = slow.Next
			if fast.Next != nil {
					fast = fast.Next.Next    
			}else {
					return false
			}
			
			if slow == fast {
					return true
			}
			// fmt.Println(fast)
	}
	
	return false
}

File: leetcode-solutions/python/141. Linked List Cycle.py
# https://leetcode.com/problems/linked-list-cycle/submissions/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return False
        
        fast = head.next
        slow = head
        
        while fast != None and fast.next != None:
            if fast == slow:
                return True
            fast = fast.next.next
            slow = slow.next

        return False
